Create ESP from charges

shehan807 opened the pull request:
This feature introduces a function, `charges_to_esp`, that performs the reverse of the RESP process — given a set of atomic charges, it computes the electrostatic potential (ESP). The implementation is quite simple, as shown in the `resp/examples/pyrazole/mpfit` directory:

```python 
import psi4
import resp

mol = psi4.geometry(""" 
 N   6.27981210  -4.25303372   0.19313865
 N   6.06870777  -2.28125697   1.80544317
 H   3.66871156  -5.64628941  -1.96684001
 H   0.69034521  -2.37545694  -0.42864236
 H   3.35917442   0.17650601   3.09265017
 H   7.35266782  -1.99461311   2.90680702
 C   3.77952586  -1.21508453   1.76858172
 C   2.38242998  -2.55505538   0.04607681
 C   4.01771468  -4.38410490  -0.88761863
""")

mol.update_geometry()

options = {'VDW_SCALE_FACTORS'  : [1.4, 1.6, 1.8, 2.0],
           'VDW_POINT_DENSITY'  : 1.0,
           'RESP_A'             : 0.0005,
           'RESP_B'             : 0.1,
           }

# MPFIT-generated charges 
charges = [-0.39242, -0.02616, 0.10538, 0.13998, 0.12150, 0.25454, 0.00422, -0.30743, 0.10040]

# Generate ESP grid files from charges
resp.charges_to_esp(mol, charges, options)

print(f"Generated grid files:")
print(f"  1_{mol.name()}_grid.dat")
print(f"  1_{mol.name()}_grid_esp.dat")
```

shehan807 committed:
created quick esp.sh for .dat -> mol2 file conversion

shehan807 committed:
added functionality to go from charges to ESP

shehan807 committed:
cleaned up dir, added comparison helper function

shehan807 committed:
updated utils with library of comparison functions

alenaizan commented:
Thank you for this pull request. I am not inclined to merge the code changes to  `resp/resp` because it is not part of the RESP method, but I see how some of these functions can be useful. I would suggest that you consolidate the code into an `examples/mpfit` directory and include only one representative example. For the example, read the RESP-generated grid, compute the classical electrostatic potential on that grid, then calculate some metrics. No need for the visualization code.

I will close this PR. Feel free to open it again if you want to contribute a simpler example.

shehan807 committed:
updated pyrazine and acetic acid examples

shehan807 committed:
removed visualization scripts

shehan807 committed:
simplified new examples

shehan807 committed:
removed example5 dir, updated readme

shehan807 committed:
simplified example

shehan807 committed:
cleanup resp

shehan807 commented:
@alenaizan, thanks for the feedback. I've updated the fork to include a simplified `example5.py` and removed the visualization scripts. Although I don't think I have permission to reopen this PR, should I create a new one?

alenaizan commented:
I just reopened the PR. Please push your latest changes and I will review the code again.

shehan807 committed:
simplified create_esp via example5.py + utils.py in example dir

shehan807 committed:
removed original resp/resp changes

shehan807 committed:
gitignore

shehan807 committed:
updated example5.py and utils.py

shehan807 commented:
@alenaizan I've updated the PR with `example5.py` and `utils.py`, in which I've included 3 essential functions, `charges_to_esp`, `compare_grid_esp`, and `create_difference_esp`, as well as any documentation and comments for their use.

alenaizan commented:
You need to do a few changes:
* Just read the RESP grid file directly and use it to calculate the electrostatic potential. You don't need to recreate the grid in the `charges_to_esp` function.
* Use the RESP charges instead of your custom MPFit charges.
* Delete the `create_difference_esp` function.

The goal of the example would be to compare the quantum electrostatic potential computed with Psi4 to the electrostatic potential generated by the RESP charges.

shehan807 committed:
reading RESP grid directly; using RESP charges instead of MPFIT; deleted create_difference_esp function

shehan807 commented:
I've made the relevant changes, thanks.

alenaizan commented:
You do not read the grid. You are just reading the ESP file. You still generate the grid manually in the `charges_to_esp` function.

You need to read grid.dat file.

shehan807 committed:
read RESP grid directly

shehan807 commented:
I see, let me know if this isn't what you had in mind from my most recent commit: 

```python 
    # Read grid points
    points = np.loadtxt(options['GRID'][0])
    if 'Bohr' in str(molecule.units()):
        points *= bohr_to_angstrom
```

alenaizan commented on examples/README.md:
```diff
@@ -3,4 +3,5 @@
 - `example2.py`: Two-stage fitting for two ethanol conformers
 - `example3.py`: Two-stage fitting for three ethanol conformers
 - `example4.py`: Two-stage fitting for four ethane-1,2-diol conformers with complex constraints
+- `example5.py`: One-stage fitting (ESP) compared to user-provided charges
```
Update description here. This is a two-stage fitting and the charges are RESP charges.

shehan807 replied:
done

alenaizan commented on examples/utils.py:
```diff
@@ -0,0 +1,137 @@
+    None
+        Writes grid.dat and grid_esp.dat files
+    """
+    if options is None:
```
Delete the lines 32-56. You don't need these lines because you are reading the grid directly.

shehan807 replied:
done (note I changed the inputs behavior slightly to simplify it)

alenaizan reviewed (changes requested):
You still need to delete the parts related to the grid generation.

shehan807 committed:
updated comments, removed lines 32-56 except for options line

shehan807 committed:
removed tmp outputs

shehan807 committed:
updated function comments and simplified inputs

alenaizan commented:
Thank you for this example.
